### PR TITLE
Add simple reflective .Data(prefix, object) capability to our fluent logger.

### PR DESCRIPTION
**Pull request by zerkz**
Includes unit tests.

This extension allows us to add Data() properties by reflectively looping through simple properties. Because I'm lazy and "debug" logging shouldn't be so painful.

Hard coded limit of 30 properties per object mapped, this number can easily change, and I plan to let the caller control this.

Will not transverse complicated objects (this is shallow reflection).

### CODE REVIEW

File: src/uShip.Logging.Tests/FluentLoggerDataBuilderTestscs.cs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using FluentAssertions;
+using NUnit.Framework;
+
+namespace uShip.Logging.Tests
+{
+    [TestFixture]
+    public class FluentLoggerDataBuilderTestscs
```
Comment (EdVinyard) [inline src/uShip.Logging.Tests/FluentLoggerDataBuilderTestscs.cs:12]: Is this name just an accident?  The trailing "cs" in the class name is unusual.

File: src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs [hunk 2]
```diff
@@ -82,6 +84,104 @@ public IFluentLoggerWriter Exception(Exception exception)
                 return this;
             }
 
+            /// <summary>
+            /// Reflectively add non-complex properties (exceptions like DateTime and DateTimeOffset) from an object to the Data dictionary to be included in the logger's message.
+            /// </summary>
+            /// <param name="prefix">Prefix that will be prepended to the property key names in the data dictionary.</param>
```
Comment (EdVinyard) [inline src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs:90]: An example here of how the prefix relates to the final name in the log would be nice.  E.g., given the prefix "foo", the value of `value.Bar` would be added with key "foo.Bar".

File: src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs [hunk 3]
```diff
@@ -82,6 +84,104 @@ public IFluentLoggerWriter Exception(Exception exception)
                 return this;
             }
 
+            /// <summary>
+            /// Reflectively add non-complex properties (exceptions like DateTime and DateTimeOffset) from an object to the Data dictionary to be included in the logger's message.
```
Comment (EdVinyard) [inline src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs:88]: Listing off the supported types explicitly here is probably a good idea, so that a client author need not read this class to know what to expect.

File: src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs [hunk 4]
```diff
@@ -82,6 +84,104 @@ public IFluentLoggerWriter Exception(Exception exception)
...
+                return Data(name, value);
+            }
+
+            /// <summary>
+            /// Recursive method that adds properties of an object to the Data dictionary to be included in the logger's message.
+            /// </summary>
+            /// <param name="name"></param>
+            /// <param name="value"></param>
+            /// <param name="topLevel"></param>
+            private void MapDataFromObject(string name, object value, bool topLevel = false)
```
Comment (EdVinyard) [inline src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs:116]: This name is a little cryptic, but your doc comment already suggests a more descriptive name: "AddPropertiesToDataDictionary".

File: src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs [hunk 5]
```diff
@@ -82,6 +84,104 @@ public IFluentLoggerWriter Exception(Exception exception)
...
+            public IFluentLoggerWriter Data(object value)
+            {
+                var name = value.GetType().Name + '_';
+                return Data(name, value);
+            }
+
+            /// <summary>
+            /// Recursive method that adds properties of an object to the Data dictionary to be included in the logger's message.
+            /// </summary>
+            /// <param name="name"></param>
```
Comment (EdVinyard) [inline src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs:113]: If you don't add any documentation to these <param> elements, just go ahead and delete them.

File: src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs (hunk 4 above)
Comment (EdVinyard) [inline src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs]: It appears that the `name` parameter is actually the `prefix` mentioned above.  I'd suggest renaming it to match.

File: src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs [hunk 6]
```diff
@@ -82,6 +84,104 @@ public IFluentLoggerWriter Exception(Exception exception)
...
+                        var propertyPrefixedName = name + prop.Name;
+                        //this seems really odd, but it is how you peel the properties's values off of the topLevel object. 
+                        var val = prop.GetValue(value, null);
+                        MapDataFromObject(propertyPrefixedName, val);
+                    }
+                    //we infer that this is definitely not a primitive and exit early. 
+                    return;
+                }
+                                
+                if (type.IsPrimitive
```
Comment (EdVinyard) [inline src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs:135]: Why limit this to primitives?  What if a complex type property has a good `ToString()` method (as they should)?  Is there a benefit in adding strongly typed info to the log message in this case?  (There really could be, but I've never used it.  You and @Eibwen may have gotten much fancier with this than I have.)

File: src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs (hunk 4 above)
Comment (EdVinyard) [inline src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs:116]: Is it permissible to pass null/empty/whitespace values for any of these input parameters?  I think you're probably missing at least some precondition checks here.

File: src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs [hunk 7]
```diff
@@ -82,6 +84,104 @@ public IFluentLoggerWriter Exception(Exception exception)
...
+                MapDataFromObject(prefix, value, true);
+                return this;
+            }
+
+            /// <summary>
+            /// Short-hand version of <see cref="Data(string,object)"/ of logging data from an object.
+            /// </summary>
+            /// <param name="value">Object's whose type will be used as the prefix to logging key name for all its properties.</param>
+            /// <returns></returns>
+            public IFluentLoggerWriter Data(object value)
```
Comment (EdVinyard) [inline src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs:104]: This is a little scary/unpredictable.  What happens if I do this twice with objects of the same type?

File: src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs [hunk 8]
```diff
@@ -82,6 +84,104 @@ public IFluentLoggerWriter Exception(Exception exception)
                 return this;
             }
 
+            /// <summary>
+            /// Reflectively add non-complex properties (exceptions like DateTime and DateTimeOffset) from an object to the Data dictionary to be included in the logger's message.
+            /// </summary>
+            /// <param name="prefix">Prefix that will be prepended to the property key names in the data dictionary.</param>
+            /// <param name="value">Object/class that we are logging properties from.</param>
+            /// <returns></returns>
+            public IFluentLoggerWriter Data(string prefix, object value)
```
Comment (EdVinyard) [inline src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs:93]: Is it OK for `prefix` to be `null`?  Empty?  Whitespace?  As a client author, I'd assume it's OK for value to be `null`, but I think that throws `NullReferenceException`.

File: src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs [hunk 9]
```diff
@@ -21,7 +23,7 @@ private class FluentLogDataBuilder : IFluentLoggerWriter
             private Exception _exception;
             private string _sql;
             private IEnumerable<KeyValuePair<string, object>> _sqlParameters;
-            private readonly IDictionary<string, object> _data = new Dictionary<string, object>();
+            internal readonly IDictionary<string, object> _data = new Dictionary<string, object>();
```
Comment (EdVinyard) [inline src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs:26]: I'd prefer you expose an internal property that returns an `IReadOnlyDictionary<string, object>` which you label as "for unit testing purposes only".  That would ultimately be safe to expose without caveat, whereas this isn't.

https://msdn.microsoft.com/en-us/library/hh136548(v=vs.110).aspx

File: src/uShip.Logging.Tests/FluentLoggerDataBuilderTestscs.cs [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+                ComplexClass = new
+                {
+                    NestedProperty = "Oh no!"
+                },
+                DateTimeProp = testDate,
+                DateTimeOffsetProp = testDateOffset
+            };
+            ClassUnderTest.Data(obj);
+            var objName = obj.GetType().Name + "_";
+            Assert.AreEqual(ClassUnderTest._data[objName + "StringProp"], testString);
```
Comment (EdVinyard) [inline src/uShip.Logging.Tests/FluentLoggerDataBuilderTestscs.cs:109]: See my comments below about the `_data` field.  I think you can get the white-box testing effect you want without this much exposure.

File: src/uShip.Logging.Tests/FluentLoggerDataBuilderTestscs.cs [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+        {
+            string testString = "test";
+            int testInt = 32;
+            long testLong = 64L;
+            double testDouble = (double)2.0;
+            float testFloat = (float)3.0;
+            DateTime testDate = new DateTime(66, 6, 6);
+            DateTimeOffset testDateOffset = new DateTimeOffset(testDate, TimeSpan.FromHours(2));
+
+            var obj = new
```
Comment (EdVinyard) [inline src/uShip.Logging.Tests/FluentLoggerDataBuilderTestscs.cs:93]: Clever use of anonymous types!  I wouldn't have thought of this, and would have written test types for everything.

**Comment by dpoindexter** [comment]
Someone with more C# knowledge than me will have to answer, but...

Is there a potential perf footgun here? We generally think of logging calls as trivially cheap (or fire-and-forget). Reflection isn't the cheapest thing, and while I'm guessing it would take a lot to become a significant source of slowness, it might be easy to accumulate reflecting calls throughout the call stack, over time.

File: src/uShip.Logging.Tests/FluentLoggerDataBuilderTestscs.cs [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+                LongProp = testLong,
+                DoubleProp = testDouble,
+                FloatProp = testFloat,
+                ComplexClass = new {
+                    NestedProperty = "Oh no!"
+                }
+            };
+            ClassUnderTest.Data(obj);
+            var objName = obj.GetType().Name + "_";
+            Assert.AreEqual(ClassUnderTest._data[objName + "StringProp"], testString);
```
Comment (EdVinyard) [inline src/uShip.Logging.Tests/FluentLoggerDataBuilderTestscs.cs:72]: Is it necessary to re-test the behavior with primitive properties here?  I can see why you need to prove that there are only the expected keys in the dictionary, but any change to primitive type behavior will probably break this test too, despite its name.

File: src/uShip.Logging.Tests/FluentLoggerDataBuilderTestscs.cs [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+                {
+                    NestedProperty = "Oh no!"
+                },
+                DateTimeProp = testDate,
+                DateTimeOffsetProp = testDateOffset
+            };
+            ClassUnderTest.Data(obj);
+            var objName = obj.GetType().Name + "_";
+            Assert.AreEqual(ClassUnderTest._data[objName + "StringProp"], testString);
+            Assert.AreEqual(ClassUnderTest._data[objName + "IntegerProp"], testInt);
```
Comment (EdVinyard) [inline src/uShip.Logging.Tests/FluentLoggerDataBuilderTestscs.cs:110]: Again, I think it would be best to keep the test focused on just one thing at a time.  Repeating the primitive test here makes all of these tests somewhat interdependent.

**Review by EdVinyard — CHANGES_REQUESTED** [review]
The three big things I'd like to see massaged a little here are:
- more complete doc comments so that client authors need never read this library (I know that's not true everywhere else, but going forward there's every reason to stick to this standard)
- more care about edge cases; should public methods throw or silently fail?
- don't expose `_data` Dictionary in implementation

File: src/uShip.Logging.Tests/FluentLoggerDataBuilderTestscs.cs (hunk 1 above)
Comment (zerkz) [reply src/uShip.Logging.Tests/FluentLoggerDataBuilderTestscs.cs:12]: Yeppppp haha. nice catch.

File: src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs (hunk 6 above)
Comment (zerkz) [reply src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs:135]: This is a great point. As long as we rely on the topLevel construct/flow, we don't necessarily need to check out types here.

**Comment by zerkz** [comment]
@dpoindexter fair point. I think the main case here is ability to debug something fast. It's pretty rare logging using this method would stay in production for a long time, but its possible....

I propose we either:
1) Determine that this kind of reflection is "cheap" enough for us for the benefit it gives the programmer.
2) Make the implementation only work in non-prod environments.
3) Scrap the idea all together.

File: src/uShip.Logging.Tests/FluentLoggerDataBuilderTestscs.cs (hunk 13 above)
Comment (zerkz) [reply src/uShip.Logging.Tests/FluentLoggerDataBuilderTestscs.cs:110]: Fair point, I was thinking about this last night, and I kind of stumbled around what if there was a bug around analyzing combinations of types. but I can easily just add another test explicitly for that.

File: src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs (hunk 9 above)
Comment (zerkz) [reply src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs:26]: will do!

File: src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs (hunk 8 above)
Comment (zerkz) [reply src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs:93]: Just tested null + string.. in concatenation null behaves the same way as an empty string.
